src/plugins/dashboard/public/application/components/dashboard_top_nav/dashboard_top_nav.tsx

```diff
@@ -53,9 +55,97 @@ const TopNav = ({
 
   const { services } = useOpenSearchDashboards<DashboardServices>();
   const { TopNavMenu, HeaderControl } = services.navigation.ui;
-  const { dashboardConfig, setHeaderActionMenu } = services;
+  const { dashboardConfig, setHeaderActionMenu, keyboardShortcut } = services;
   const { setAppRightControls } = services.application;
 
+  // Get nav actions for direct function calls
+  const keyboardNavActions = getNavActions(
+    appState,
+    savedDashboardInstance,
+    services,
+    dashboard,
+    dashboardIdFromUrl,
+    currentContainer
+  );
+
+  const handleToggleDashboardEdit = useCallback(() => {
+    const isEditMode = currentAppState?.viewMode === ViewMode.EDIT;
+    const actionId = isEditMode ? TopNavIds.EXIT_EDIT_MODE : TopNavIds.ENTER_EDIT_MODE;
+    if (keyboardNavActions && keyboardNavActions[actionId]) {
+      keyboardNavActions[actionId]();
+    }
+  }, [keyboardNavActions, currentAppState]);
+
+  const handleSave = useCallback(() => {
+    if (keyboardNavActions && keyboardNavActions[TopNavIds.SAVE]) {
+      keyboardNavActions[TopNavIds.SAVE]();
+    }
+  }, [keyboardNavActions]);
+
+  const handleAdd = useCallback(() => {
+    // Note: Cannot use navActions[TopNavIds.ADD_EXISTING]() directly because it requires
+    // an anchorElement parameter for popover positioning. Keyboard shortcuts have no
+    // anchor element, so we simulate clicking the actual button to provide proper context.
+    const addButton = document.querySelector(DASHBOARD_DOM_SELECTORS.ADD_PANEL_BUTTON);
+    if (addButton && !addButton.hasAttribute('disabled')) {
+      (addButton as HTMLElement).click();
+    }
+  }, []);
+
+  // Register/unregister save shortcut based on edit mode
+  useEffect(() => {
+    if (currentAppState?.viewMode === 'edit' && keyboardShortcut) {
+      keyboardShortcut.register({
+        id: 'save_dashboard',
+        pluginId: 'dashboard',
+        name: 'Save Dashboard',
+        category: 'editing / save',
+        keys: 'cmd+s',
+        execute: handleSave,
+      });
+
+      // Cleanup: unregister when leaving edit mode or component unmounts
+      return () => {
+        keyboardShortcut.unregister({
+          id: 'save_dashboard',
+          pluginId: 'dashboard',
+        });
+      };
+    }
+  }, [currentAppState?.viewMode, keyboardShortcut, handleSave]);
+
+  // Register/unregister add shortcut based on edit mode
+  useEffect(() => {
+    if (currentAppState?.viewMode === 'edit' && keyboardShortcut) {
+      keyboardShortcut.register({
+        id: 'add_dashboard',
```

Comment:
this id is confusing it's not really adding a dashboard it's a dding a panel or a visualization